go-controller/pkg/node/nftables/util.go

```diff
@@ -5,39 +5,147 @@ package nftables
+// are returned for objects that don't exist.
+//
+// Currently DeleteObjects only supports knftables.Element. Every Element in objects will
+// be deleted from its corresponding set/map (if it is currently a member). If an Element
+// refers to a set/map that doesn't already exist, an error will be returned.
+//
+// If you pass map elements without the Value field set then DeleteObjects will need to
+// first "list" the map to find its current contents, to ensure that it only tries to
+// delete the elements that are actually present. (With set elements, or map elements with
+// Value set, it can just do an Add+Delete to get the effect of "delete whether or not it
+// previously existed".)
+func DeleteObjects(ctx context.Context, objects []knftables.Object) error {
 	nft, err := GetNFTablesHelper()
 	if err != nil {
 		return err
 	}
 
+	// If there are any partial Map Elements, list their existing contents
+	existingMaps := make(map[string][]*knftables.Element)
+	for _, obj := range objects {
+		switch typed := obj.(type) {
+		case *knftables.Element:
+			if typed.Map != "" && len(typed.Value) == 0 && existingMaps[typed.Map] == nil {
+				elements, err := nft.ListElements(ctx, "map", typed.Map)
+				if err != nil {
+					return err
+				}
+				existingMaps[typed.Map] = elements
+			}
+		default:
+			return fmt.Errorf("unsupported object type %T passed to DeleteObjects", obj)
+		}
+	}
+
+	// Now build the actual transaction
 	tx := nft.NewTransaction()
+	for _, obj := range objects {
+		switch typed := obj.(type) {
+		case *knftables.Element:
+			if typed.Map != "" && len(typed.Value) == 0 {
+				// We can't tx.Add() a Map Element with no Value, so try
+				// to find its existing value in the List output from
+				// above. If the element doesn't appear in that output
+				// then we just skip trying to delete it. Otherwise, we do
+				// the Add+Delete below just like in the normal case; the
+				// Add *should* be a no-op, but doing it anyway makes this
+				// work right even if another thread Deletes the element
+				// before we get to it (as long as they don't add it back
+				// with a different value).
+				typed = findElement(existingMaps[typed.Map], typed.Key)
+				if typed == nil {
+					continue
+				}
+			}
+
+			// Do Add+Delete, which ensures the object is deleted whether or
+			// not it previously existed.
+			tx.Add(typed)
+			tx.Delete(typed)
+		default:
+		}
+	}
+	return nft.Run(ctx, tx)
+}
+
+func findElement(elements []*knftables.Element, key []string) *knftables.Element {
+elemLoop:
 	for _, elem := range elements {
-		// We add+delete the elements, rather than just deleting them, so that if
-		// they weren't already in the set/map, we won't get an error on delete.
-		tx.Add(elem)
-		tx.Delete(elem)
+		if len(elem.Key) != len(key) {
+			// All elements have the same key length, so if one fails, they all fail.
+			return nil
+		}
+		for i := range elem.Key {
+			if elem.Key[i] != key[i] {
+				continue elemLoop
```

Comment:
i am not following this. Why do you jump to back to the start again?

Reply:
"if this element doesn't have the correct key value, then continue on to the next element."

lines 102-107 could be rewritten as

    if reflect.DeepEqual(elem.Key, key) {
            return elem
    }

(except `reflect.DeepEqual` is inefficient)

Reply:
makes sense - brain fart remembering how `range` worked with this `continue` to `elemLoop`